### ISSUE TITLE
allow input as a matrix of posterior draws, N, and a log-likelihood function

### ISSUE THREAD

**Issue by bgoodri** [jgabry/loo#8]
You could save a lot of RAM if the user did not have to input a SxN matrix of log-likelihoods. I think it should be possible for the user to pass a function like
```
ll <- function(i, draws) {
  # return a vector of log-likelihoods for observation i 
}
```
and have functions like `loo` call `vapply(1:N, FUN = foo, FUN.VALUE = double(?))` where `foo` is defined in loo's NAMESPACE and calls `ll(i,draws)` internally. Then `foo(i)` takes the result of `ll(i,draws)` and returns whatever quantities are needed from the i-th observation. This would only require memory to hold about S doubles at a time.

**Comment by jgabry** [comment]
Thanks Ben! Yeah, I actually have two different versions that I've been
playing with that do something to that. I've been profiling using models
with various N to get a better sense for this since, as you point
out,  memory is definitely an issue for models with large N. It's not too
bad even for N on the order of 10^3, but beyond that it's a concern.

The interface is much more intuitive with the log lik matrix, as I think
some users would find it strange to pass a function like this (it's not too
common). But I definitely like the idea. Perhaps for the next version of
the package we can allow either input (matrix or function) and encourage
users with big models (in terms of N, not S) to avoid the matrix. What do
you think?

**Comment by bgoodri** [comment]
Yes, it should be an option in addition to the giant matrix.

**Comment by jgabry** [comment]
When I get that version tested and up on GitHub I'll let you know so you
can play with it and give some feedback, if you're amenable.

**Comment by bgoodri** [comment]
This is still not quite enough to make it possible to call `loo` for moderate sized datasets. It doesn't do us any good to avoid calculating the SxN matrix of log-likelihoods if we have to store a SxN matrix of linear predictors or expectations to do so. Within `loo:::psislw`,  the `.psis_loop` function should instead call
```
ll_i <- llfun(i = i, data = llargs$data[i,,drop=FALSE], draws = llargs$draws)
```
when `LL_FUN` is `TRUE`. Then, the caller's `llfun` is responsible calculating the linear predictor, expectation, and ultimately the log-likelihood for the i-th observation over the S draws in the posterior. This way, only S numbers are briefly kept in memory which are freed when `.psis_loop` returns two scalars.

**Comment by jgabry** [comment]
@bgoodri good point. I'll make that change.

**Comment by jgabry** [comment]
@bgoodri Ok, I hope/think this should be fixed by [a7d08d7](https://github.com/jgabry/loo/commit/a7d08d7af4308a0def9261d5e911b269397d148c). If that looks good to you I'll finalize it and start getting it ready for a release.

**Comment by bgoodri** [comment]
I think the `data` argument could be a data.frame as an alternative to a
matrix. Anyway, let's get it working again on the rstanarm side. I don't
think there is any need to rush a loo release just for this feature if
rstanarm is still a few weeks away.
